iceoryx2-bb/log/src/logger/file.rs

```diff
@@ -61,78 +67,43 @@ impl Debug for Entry {
     }
 }
 
-struct State {
-    buffer: VecDeque<Entry>,
-    keep_running: bool,
-}
-
-impl Default for State {
-    fn default() -> Self {
-        Self {
-            buffer: VecDeque::new(),
-            keep_running: true,
-        }
-    }
-}
-
 /// A logger that logs all messages into a file. It implements an active object pattern. A
 /// background thread waits on a queue of log messages and whenever a new message is added.
 pub struct Logger {
-    state: Arc<Mutex<State>>,
-    trigger: Arc<Condvar>,
+    sender: Arc<Sender<Message>>,
     start_time: Instant,
     _background_thread: Arc<Option<JoinHandle<()>>>,
 }
 
 impl Logger {
     /// Creates a new file logger.
     pub fn new(file_name: &str) -> Self {
-        let state = Arc::new(Mutex::new(State::default()));
-        let trigger = Arc::new(Condvar::new());
         let mut file = OpenOptions::new()
             .append(true)
             .create(true)
             .open(file_name)
             .expect("Open log file for writing.");
 
-        let self_state = state.clone();
-        let self_trigger = trigger.clone();
+        let (sender, receiver) = channel();
 
         let write_buffer_to_file = move || loop {
-            let lock = state.lock().expect("Acquire internal state mutex.");
-
-            let lock = trigger
-                .wait_while(lock, |state: &mut State| {
-                    state.buffer.is_empty() || !state.keep_running
-                })
-                .expect("Wait for internal state trigger.");
-            drop(lock);
-
-            loop {
-                // acquiring the lock only to get the next buffer to minimize contention as much
-                // as possible.
-                let mut lock = state.lock().expect("Acquire internal state mutex.");
-                let buffer = lock.buffer.pop_front();
-                drop(lock);
-
-                match buffer {
-                    Some(entry) => file
-                        .write_all(format!("{:?}\n", entry).as_bytes())
-                        .expect("Writing log message into log file."),
-                    None => break,
-                }
-            }
+            match receiver.recv() {
+                Ok(Message::Entry(entry)) => file
+                    .write_all(format!("{:?}\n", entry).as_bytes())
+                    .expect("Writing log message into log file."),
+                Ok(Message::Stop) => break,
+                Err(e) => file
+                    .write_all(
+                        format!("[This should never happen!] File Logger got error: {:?}", e)
+                            .as_bytes(),
+                    )
+                    .expect("Write log message into log file."),
+            };
             file.sync_all().expect("Sync log file with disc.");
```

Comment:
will `sync_all` cost a lot for every piece of messages? I have searched the forum mentioned it [in this post](https://users.rust-lang.org/t/patterns-for-flush-sync-in-async-i-o/101563). Hdyt? 

> B) Asking the OS to dump its in-memory ... . it can be a slow and expensive operation.
> This operation is not required for data to become visible to other applications.

Reply:
I guess the idea was to have the logs available when a crash occurs. Maybe this can be done via a crash handler. On the other side, AFAIK there are no logs in the hot path, so it should also not affect performance.

Reply:
This is why all of this is handled in a background thread. The hot path just uses the channel to move the log message out of the hot thread into the log thread.

Reply:
I think the worst case scenario is, when a crash happens and the logger didn't sync the file and therefore the final log messages are not written.

However, I also want to highlight that this does not need to be the final implementation. If we realize this becomes a bottleneck we can put on our thinking caps and try harder.